### ISSUE TITLE
scrub plugin: preserve tags

### ISSUE THREAD

**Issue by m-urban** [sampsyo/beets#1514]
I just happened to run into the following issue: One of my plugins taps into the `write` event to [preserve the file's year tag](https://github.com/sampsyo/beets/issues/1498). In the event handler, I read the file's tags again, constructing a `MediaFile` object. While `import`ing new songs, I noticed that the properties of those MediaFiles were returning `None` (e.g. `mf.year == None`), although they should contain valid values. Turns out that disabling the `scrub` plugin fixes this.
Now what I think is happening here is that the `scrub` plugin comes into action and deletes the file's meta data right before my other plugin kicks in, only to find empty tags.

IMO there are two ways to fix this:
- Make the execution order of plugins configurable (probably a smart idea anyways, but could be a big chunk of work)
- Use a white list in `scrub`, permitting to preserve listed tags. Something along the lines of:
```
scrub:
    auto: yes
    preserve: art comments year
```
I am not sure about the former, but the latter seems not too difficult to achieve. Similar to what we do with the cover art on a manual `scrub` operation: Save listed tags, clear tag frame, restore preserved tags. This would be a more generalized approach to solving #571.

**Comment by sampsyo** [comment]
Hmm—I understand the issue, which sounds complex, but I'm a little uncertain about the level at which we're aiming to solve this.

Scrub has a pretty straightforward purpose: make the files' metadata match the beets database, no matter what. Walking that back makes the story a bit more complicated.

It sounds like the real issue you're having here, described elsewhere, is that beets wants to homogenize the year across an entire album. It seems like the "right" solution is to address that directly: to provide a way to disable that album-matching effect when you don't want it. Your current strategy, where you leave the on-disk metadata intact, is a totally reasonable workaround, but it's not really that ideal solution—so I'm not sure we should work on changing other plugins to enshrine that hack as the right way to do it.

Is there any chance we can get a more general, long-term solution by cleaning up the functionality of `update`, and anywhere else that's giving you trouble with your per-track release years?

**Comment by m-urban** [comment]
I totally see your point and need to agree with you - workarounds to enable workarounds are probably not the greatest approach in the world. However, my approach shouldn't effect existing functionality

But let me clarify my problem: I need to store the track_year in the database. Flexible attributes seem to be the ideal fit for that. The issue is, I need to *read* the file's tags at some point during the `import` process. The only hook I can use is the `write` event, which gives me access to the file's path and consequently to its metadata. Now at this point in time, other plugins like scrub, which listen to the same event, can have kicked in already and damage the tags.

This is what happens with scrub: it removes tags entirely before another plugin gets a chance to read the integral source file (e.g. to save its metadata to flexattrs). This is actually unrelated to my specific complex use case, but can generally effect any plugin.

**Comment by sampsyo** [comment]
Yes, good point: it seems like the right general solution might be to let you map the right data to your preferred flexible attribute in a robust way. Maybe we can figure out what the right hook (on both reads and writes) would be to let a plugin implement that cleanly.

**Comment by m-urban** [comment]
I think we should introduce something like a `before_write` event that 
passes a path reference and a library item. We can fire the event right 
after creating the database entry and just before modifying the file's 
metatags in `write`.
We can probably not programatically enforce it, but we should adjust the 
documentation to state that plugins should not tamper with imported 
files before the `write` hook to ensure files' integrity. This would 
give plugins a robust chance to read valid meta data.

However, I would still need to find a way to map my library's track_year 
flexible attribute to the file's year tag on possibly every write. I 
can't just rewrite the tag and assign the track_year to the 
`tags['year']` entry, because of the same reason - I cannot guarantee 
that scrub (or virtually any other plugin) isn't replacing it again with 
some garbage tag, just because it is executed after my own plugin.
